std/parallelism.d

```diff
@@ -4303,243 +4321,311 @@ version(parallelismStressTest)
-            {
-                return sqrt(num);
-            }
+    // Make sure work is reasonably balanced among threads.  This test is
+    // non-deterministic and is more of a sanity check than something that
+    // has an absolute pass/fail.
+    shared(uint)[void*] nJobsByThread;
+    foreach (thread; poolInstance.pool)
+    {
+        nJobsByThread[cast(void*) thread] = 0;
+    }
+    nJobsByThread[ cast(void*) Thread.getThis()] = 0;
 
-            static real[] parallelSqrt(real[] nums)
-            {
-                return poolInstance.amap!mySqrt(nums);
-            }
+    foreach (i; poolInstance.parallel( iota(0, 1_000_000), 100 ))
+    {
+        atomicOp!"+="( nJobsByThread[ cast(void*) Thread.getThis() ], 1);
+    }
 
-            real[][] sqrtMatrix = poolInstance.amap!parallelSqrt(matrix);
+    version(parallelismStressTest) stderr.writeln("\nCurrent thread is:  ",
+    cast(void*) Thread.getThis());
+    version(parallelismStressTest) stderr.writeln("Workload distribution:  ");
 
-            foreach (i, row; sqrtMatrix)
-            {
-                foreach (j, elem; row)
-                {
-                    real shouldBe = sqrt( cast(real) i * j);
-                    assert(approxEqual(shouldBe, elem));
-                    sqrtMatrix[i][j] = shouldBe;
-                }
-            }
+    version(parallelismStressTest)
+    {
+        foreach (k, v; nJobsByThread)
+        {
+            stderr.writeln(k, '\t', v);
+        }
+    }
+    else
+    {
+        foreach (k, v; nJobsByThread)
+        {
+        }
+    }
 
-            auto saySuccess = task(
-            {
-                stderr.writeln(
-                    "Success doing matrix stuff that involves nested pool use.");
-            });
-            poolInstance.put(saySuccess);
-            saySuccess.workForce();
+    // Test whether amap can be nested.
+    real[][] matrix = new real[][](1000, 1000);
+    foreach (i; poolInstance.parallel( iota(0, matrix.length) ))
+    {
+        foreach (j; poolInstance.parallel( iota(0, matrix[0].length) ))
+        {
+            matrix[i][j] = i * j;
+        }
+    }
 
-            // A more thorough test of amap, reduce:  Find the sum of the square roots of
-            // matrix.
+    static real[] parallelSqrt(real[] nums)
+    {
+        return poolInstance.amap!(mySqrt)(nums);
+    }
 
-            static real parallelSum(real[] input)
-            {
-                return poolInstance.reduce!"a + b"(input);
-            }
+    static real parallelSum(real[] input)
+    {
+        return poolInstance.reduce!"a + b"(input);
+    }
 
-            auto sumSqrt = poolInstance.reduce!"a + b"(
-                               poolInstance.amap!parallelSum(
-                                   sqrtMatrix
-                               )
-                           );
+    // Get around weird bugs having to do w/ sqrt being an intrinsic:
+    real[][] sqrtMatrix = poolInstance.amap!parallelSqrt(matrix);
```

Comment:
amap can not call parallelSqrt that that local function was trying to access poolInstance which is located on the same stack